File: packages/validation/__tests__/plugin.test.ts
import { describe, it, expect } from 'vitest';
import { validateExtension } from '../src/index.js';
import { createZip } from './helpers.js';

describe('Plugin Validation', () => {
  it('passes for a valid minimal plugin with skills', async () => {
    const zip = createZip({
      '.claude-plugin/plugin.json': JSON.stringify({
        name: 'my-plugin',
        version: '1.0.0',
        description: 'A test plugin',
      }),
      'skills/search/SKILL.md': `---
description: Search for things
---

# Search Skill

This skill searches for things effectively and returns results.
`,
      'README.md': '# My Plugin\n\nA useful plugin.',
    });

    const result = await validateExtension('plugin', zip, 'plugin.zip');
    expect(result.valid).toBe(true);
    expect(result.metadata).not.toBeNull();
    expect(result.metadata!.name).toBe('my-plugin');
  });

  it('passes for a plugin with all component types', async () => {
    const zip = createZip({
      '.claude-plugin/plugin.json': JSON.stringify({
        name: 'full-plugin',
        version: '2.0.0',
        description: 'Full plugin',
      }),
      'skills/lint/SKILL.md': `---
description: Lint your code
---

# Lint Skill

This skill lints your code and provides feedback on issues.
`,
      'commands/deploy.sh': '#!/bin/bash\necho "deploying"',
      'agents/reviewer/config.json': '{}',
      'hooks/hooks.json': '{"pre-commit": []}',
      '.mcp.json': '{"servers": {}}',
      '.lsp.json': '{"servers": {}}',
      'README.md': '# Full Plugin\n\nEverything included.',
    });

    const result = await validateExtension('plugin', zip, 'plugin.zip');
    expect(result.valid).toBe(true);
  });

  it('fails when plugin.json is missing', async () => {
    const zip = createZip({
      'skills/search/SKILL.md': `---
description: Search
---

# Search Skill

This skill searches for things effectively.
`,
      'README.md': '# Plugin',
    });

    const result = await validateExtension('plugin', zip, 'plugin.zip');
    expect(result.valid).toBe(false);
    expect(result.errors).toContainEqual(
      expect.objectContaining({
        code: 'MISSING_FILE',
        path: '.claude-plugin/plugin.json',
      }),
    );
  });

  it('fails when plugin.json has invalid JSON', async () => {
    const zip = createZip({
      '.claude-plugin/plugin.json': 'not json {{{',
      'skills/a/SKILL.md': `---
description: test
---

# Test Skill

A test skill for validation testing purposes.
`,
      'README.md': '# Plugin',
    });

    const result = await validateExtension('plugin', zip, 'plugin.zip');
    expect(result.valid).toBe(false);
    expect(result.errors).toContainEqual(
      expect.objectContaining({ code: 'INVALID_JSON' }),
    );
  });

  it('fails when plugin name is not kebab-case', async () => {
    const zip = createZip({
      '.claude-plugin/plugin.json': JSON.stringify({
        name: 'MyPlugin',
        version: '1.0.0',
      }),
      'skills/a/SKILL.md': `---
description: test
---

# Test Skill

A test skill for validation testing purposes.
`,
      'README.md': '# Plugin',
    });

    const result = await validateExtension('plugin', zip, 'plugin.zip');
    expect(result.valid).toBe(false);
    expect(result.errors).toContainEqual(
      expect.objectContaining({
        code: 'INVALID_SCHEMA',
        message: expect.stringContaining('kebab-case'),
      }),
    );
  });

  it('fails when no components are present', async () => {
    const zip = createZip({
      '.claude-plugin/plugin.json': JSON.stringify({
        name: 'empty-plugin',
        version: '1.0.0',
      }),
      'README.md': '# Empty Plugin',
    });

    const result = await validateExtension('plugin', zip, 'plugin.zip');
    expect(result.valid).toBe(false);
    expect(result.errors).toContainEqual(
      expect.objectContaining({ code: 'NO_COMPONENTS' }),
    );
  });

  it('fails when nested skill validation fails', async () => {
    const zip = createZip({
      '.claude-plugin/plugin.json': JSON.stringify({
        name: 'bad-skills',
        version: '1.0.0',
        description: 'Plugin with bad skills',
      }),
      'skills/broken/SKILL.md': `---
name: broken-skill
---

Too short
`,
      'README.md': '# Plugin',
    });

    const result = await validateExtension('plugin', zip, 'plugin.zip');
    expect(result.valid).toBe(false);
    // Should have frontmatter error from nested skill (missing description)
    expect(result.errors.some((e) => e.code === 'INVALID_FRONTMATTER')).toBe(
      true,
    );
  });

  it('fails when hooks.json is invalid JSON', async () => {
    const zip = createZip({
      '.claude-plugin/plugin.json': JSON.stringify({
        name: 'bad-hooks',
        version: '1.0.0',
        description: 'Plugin with bad hooks',
      }),
      'hooks/hooks.json': 'not valid json',
      'README.md': '# Plugin',
    });

    const result = await validateExtension('plugin', zip, 'plugin.zip');
    expect(result.valid).toBe(false);
    expect(result.errors).toContainEqual(
      expect.objectContaining({
        code: 'INVALID_JSON',
        path: 'hooks/hooks.json',
      }),
    );
  });

  it('fails when .mcp.json is invalid JSON', async () => {
    const zip = createZip({
      '.claude-plugin/plugin.json': JSON.stringify({
        name: 'bad-mcp',
        version: '1.0.0',
        description: 'Plugin with bad mcp',
      }),
      '.mcp.json': '{{invalid',
      'README.md': '# Plugin',
    });

    const result = await validateExtension('plugin', zip, 'plugin.zip');
    expect(result.valid).toBe(false);
    expect(result.errors).toContainEqual(
      expect.objectContaining({ code: 'INVALID_JSON', path: '.mcp.json' }),
    );
  });

  it('fails when README.md is missing', async () => {
    const zip = createZip({
      '.claude-plugin/plugin.json': JSON.stringify({
        name: 'no-readme',
        version: '1.0.0',
        description: 'No readme',
      }),
      'skills/a/SKILL.md': `---
description: test skill
---

# Test Skill

A test skill for validation testing purposes.
`,
    });

    const result = await validateExtension('plugin', zip, 'plugin.zip');
    expect(result.valid).toBe(false);
    expect(result.errors).toContainEqual(
      expect.objectContaining({ code: 'MISSING_FILE', path: 'README.md' }),
    );
  });

  it('detects path traversal in plugin.json values', async () => {
    const zip = createZip({
      '.claude-plugin/plugin.json': JSON.stringify({
        name: 'evil-plugin',
        version: '1.0.0',
        description: 'Evil plugin',
        configPath: '../../etc/passwd',
      }),
      'skills/a/SKILL.md': `---
description: test skill
---

# Test Skill

A test skill for validation testing purposes.
`,
      'README.md': '# Evil Plugin',
    });

    const result = await validateExtension('plugin', zip, 'plugin.zip');
    expect(result.valid).toBe(false);
    expect(result.errors).toContainEqual(
      expect.objectContaining({ code: 'PATH_TRAVERSAL' }),
    );
  });

  it('warns when version is missing from plugin.json', async () => {
    const zip = createZip({
      '.claude-plugin/plugin.json': JSON.stringify({
        name: 'no-version',
        description: 'A plugin',
      }),
      'skills/a/SKILL.md': `---
description: test skill
---

# Test Skill

A test skill for validation testing purposes.
`,
      'README.md': '# Plugin',
    });

    const result = await validateExtension('plugin', zip, 'plugin.zip');
    expect(result.valid).toBe(true);
    expect(result.warnings).toContainEqual(
      expect.objectContaining({ code: 'MISSING_VERSION' }),
    );
  });

  it('warns when description is missing from plugin.json', async () => {
    const zip = createZip({
      '.claude-plugin/plugin.json': JSON.stringify({
        name: 'no-desc',
        version: '1.0.0',
      }),
      'skills/a/SKILL.md': `---
description: test skill
---

# Test Skill

A test skill for validation testing purposes.
`,
      'README.md': '# Plugin',
    });

    const result = await validateExtension('plugin', zip, 'plugin.zip');
    expect(result.valid).toBe(true);
    expect(result.warnings).toContainEqual(
      expect.objectContaining({ code: 'MISSING_DESCRIPTION' }),
    );
  });
});

File: cli/internal/validation/plugin_test.go
package validation

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func validPluginFiles() map[string]string {
	return map[string]string{
		".claude-plugin/plugin.json":  `{"name":"test-plugin","version":"1.0.0","description":"A test plugin"}`,
		"skills/search/SKILL.md":      "---\ndescription: Search skill\n---\nThis is the search skill body content.",
		"README.md":                   "# Test Plugin\nA test plugin.",
	}
}

func TestValidatePlugin_Valid(t *testing.T) {
	data := createZip(t, validPluginFiles())
	result := ValidateExtension(TypePlugin, data, "test.zip")
	assert.True(t, result.Valid)
	require.NotNil(t, result.Metadata)
	assert.Equal(t, "test-plugin", result.Metadata.Name)
	assert.Equal(t, "1.0.0", result.Metadata.Version)
}

func TestValidatePlugin_MissingPluginJSON(t *testing.T) {
	files := map[string]string{
		"skills/search/SKILL.md": "---\ndescription: test\n---\nThis is the skill body content.",
		"README.md":              "readme",
	}
	data := createZip(t, files)
	result := ValidateExtension(TypePlugin, data, "test.zip")
	assert.False(t, result.Valid)
	assert.Equal(t, "MISSING_FILE", result.Errors[0].Code)
}

func TestValidatePlugin_InvalidPluginJSON(t *testing.T) {
	files := validPluginFiles()
	files[".claude-plugin/plugin.json"] = "not json"
	data := createZip(t, files)
	result := ValidateExtension(TypePlugin, data, "test.zip")
	assert.False(t, result.Valid)
	assert.Equal(t, "INVALID_JSON", result.Errors[0].Code)
}

func TestValidatePlugin_InvalidName(t *testing.T) {
	files := validPluginFiles()
	files[".claude-plugin/plugin.json"] = `{"name":"InvalidName","version":"1.0.0","description":"test"}`
	data := createZip(t, files)
	result := ValidateExtension(TypePlugin, data, "test.zip")
	assert.False(t, result.Valid)
	hasErr := false
	for _, e := range result.Errors {
		if e.Code == "INVALID_SCHEMA" {
			hasErr = true
		}
	}
	assert.True(t, hasErr)
}

func TestValidatePlugin_WarningsOnMissingVersionDescription(t *testing.T) {
	files := map[string]string{
		".claude-plugin/plugin.json": `{"name":"test-plugin"}`,
		"commands/run.sh":            "#!/bin/bash\necho hello",
		"README.md":                  "readme",
	}
	data := createZip(t, files)
	result := ValidateExtension(TypePlugin, data, "test.zip")
	assert.True(t, result.Valid)
	assert.Len(t, result.Warnings, 2)

	codes := make(map[string]bool)
	for _, w := range result.Warnings {
		codes[w.Code] = true
	}
	assert.True(t, codes["MISSING_VERSION"])
	assert.True(t, codes["MISSING_DESCRIPTION"])
}

func TestValidatePlugin_NoComponents(t *testing.T) {
	files := map[string]string{
		".claude-plugin/plugin.json": `{"name":"test-plugin","version":"1.0.0","description":"test"}`,
		"README.md":                  "readme",
	}
	data := createZip(t, files)
	result := ValidateExtension(TypePlugin, data, "test.zip")
	assert.False(t, result.Valid)
	hasErr := false
	for _, e := range result.Errors {
		if e.Code == "NO_COMPONENTS" {
			hasErr = true
		}
	}
	assert.True(t, hasErr)
}

func TestValidatePlugin_WithMcpJSON(t *testing.T) {
	files := map[string]string{
		".claude-plugin/plugin.json": `{"name":"test-plugin","version":"1.0.0","description":"test"}`,
		".mcp.json":                  `{"mcpServers":{}}`,
		"README.md":                  "readme",
	}
	data := createZip(t, files)
	result := ValidateExtension(TypePlugin, data, "test.zip")
	assert.True(t, result.Valid)
}

func TestValidatePlugin_WithHooksJSON(t *testing.T) {
	files := map[string]string{
		".claude-plugin/plugin.json": `{"name":"test-plugin","version":"1.0.0","description":"test"}`,
		"hooks/hooks.json":           `{"hooks":[]}`,
		"README.md":                  "readme",
	}
	data := createZip(t, files)
	result := ValidateExtension(TypePlugin, data, "test.zip")
	assert.True(t, result.Valid)
}

func TestValidatePlugin_InvalidHooksJSON(t *testing.T) {
	files := map[string]string{
		".claude-plugin/plugin.json": `{"name":"test-plugin","version":"1.0.0","description":"test"}`,
		"hooks/hooks.json":           "not json",
		"README.md":                  "readme",
	}
	data := createZip(t, files)
	result := ValidateExtension(TypePlugin, data, "test.zip")
	assert.False(t, result.Valid)
	hasErr := false
	for _, e := range result.Errors {
		if e.Code == "INVALID_JSON" && e.Path == "hooks/hooks.json" {
			hasErr = true
		}
	}
	assert.True(t, hasErr)
}

func TestValidatePlugin_InvalidMcpJSON(t *testing.T) {
	files := map[string]string{
		".claude-plugin/plugin.json": `{"name":"test-plugin","version":"1.0.0","description":"test"}`,
		".mcp.json":                  "not json",
		"README.md":                  "readme",
	}
	data := createZip(t, files)
	result := ValidateExtension(TypePlugin, data, "test.zip")
	assert.False(t, result.Valid)
}

func TestValidatePlugin_MissingReadme(t *testing.T) {
	files := map[string]string{
		".claude-plugin/plugin.json": `{"name":"test-plugin","version":"1.0.0","description":"test"}`,
		"commands/run.sh":            "#!/bin/bash\necho hello",
	}
	data := createZip(t, files)
	result := ValidateExtension(TypePlugin, data, "test.zip")
	assert.False(t, result.Valid)
}

func TestValidatePlugin_PathTraversalInConfig(t *testing.T) {
	files := map[string]string{
		".claude-plugin/plugin.json": `{"name":"test-plugin","version":"1.0.0","description":"test","config":"../../etc/passwd"}`,
		"commands/run.sh":            "#!/bin/bash\necho hello",
		"README.md":                  "readme",
	}
	data := createZip(t, files)
	result := ValidateExtension(TypePlugin, data, "test.zip")
	assert.False(t, result.Valid)
	hasErr := false
	for _, e := range result.Errors {
		if e.Code == "PATH_TRAVERSAL" {
			hasErr = true
		}
	}
	assert.True(t, hasErr)
}

func TestValidatePlugin_WithCommandsComponent(t *testing.T) {
	files := map[string]string{
		".claude-plugin/plugin.json": `{"name":"test-plugin","version":"1.0.0","description":"test"}`,
		"commands/hello.sh":          "#!/bin/bash\necho hello",
		"README.md":                  "readme",
	}
	data := createZip(t, files)
	result := ValidateExtension(TypePlugin, data, "test.zip")
	assert.True(t, result.Valid)
}

func TestValidatePlugin_WithAgentsComponent(t *testing.T) {
	files := map[string]string{
		".claude-plugin/plugin.json": `{"name":"test-plugin","version":"1.0.0","description":"test"}`,
		"agents/agent.md":            "# Agent\nAn agent.",
		"README.md":                  "readme",
	}
	data := createZip(t, files)
	result := ValidateExtension(TypePlugin, data, "test.zip")
	assert.True(t, result.Valid)
}

func TestValidatePlugin_WithLspJSON(t *testing.T) {
	files := map[string]string{
		".claude-plugin/plugin.json": `{"name":"test-plugin","version":"1.0.0","description":"test"}`,
		".lsp.json":                  `{"servers":[]}`,
		"README.md":                  "readme",
	}
	data := createZip(t, files)
	result := ValidateExtension(TypePlugin, data, "test.zip")
	assert.True(t, result.Valid)
}

func TestValidatePlugin_WithRootDir(t *testing.T) {
	files := map[string]string{
		"myplugin/.claude-plugin/plugin.json": `{"name":"test-plugin","version":"1.0.0","description":"test"}`,
		"myplugin/commands/run.sh":            "#!/bin/bash\necho hello",
		"myplugin/README.md":                  "readme",
	}
	data := createZip(t, files)
	result := ValidateExtension(TypePlugin, data, "test.zip")
	assert.True(t, result.Valid)
}
